### PR TITLE
fix dygraph auto-gen auto_parallel code when some output is null

### PR DESCRIPTION
**Pull request by zhiqiu**
### PR Category

Auto Parallel

### PR Types

Bug fixes

### Description

The auto-generated dygraph code for auto_parallel does not handle nullptr output, so it may result in `core dump` when some grad_output  is empty because of stop_gradient or detach.
For example, the tensor `out.detach()` in concat needs no grad in backward.
```
    a = paddle.rand([4, 16, 16, 4])
    a.stop_gradient = False
    a = dist.shard_tensor(a, mesh, [dist.Replicate()])
    out, var = paddle.split(a, 2, axis = 1)
    out = paddle.concat([out.detach(), var])
    loss = paddle.mean(out)
    loss.backward()
```
Pcard-76459

### CODE REVIEW

**Review by heavyrain-lzy — APPROVED** [review]
LGTM

**Review by luotao1 — APPROVED** [review]

**Review by Aurelius84 — APPROVED** [review]
LGTM For const_cast

**Review by jeff41404 — APPROVED** [review]
